src/x/server/server.go

```diff
@@ -140,11 +146,34 @@ func (s *server) Serve(l net.Listener) error {
 	return nil
 }
 
+func (s *server) maybeUpgradeToTLS(conn SecuredConn) (SecuredConn, error) {
+	if s.tlsConfigManager.ServerMode() == xtls.Disabled {
+		return conn, nil
+	}
+	isTLSConnection, err := conn.IsTLS()
+	if err != nil {
+		return nil, err
+	}
+	if isTLSConnection {
```

Comment:
Nit: I think this might read a bit more clearly with some early returns:

```
if !isTLS && s.tlsConfigManager.ServerMode() == xtls.Enforced {
  // error
} else if !isTLS {
  return conn
}

// this is a TLS connection
// upgrade to TLS
```